gk/main.c

```diff
@@ -1175,6 +1245,384 @@ gk_conf_put(struct gk_config *gk_conf)
+			goto out;
+
+		new_fib = &ltbl->fib_tbl6[fib_id];
+
+		/*
+		 * Add the fib entry for the IPv6 address 
+		 * to the IPv6 LPM table.
+		 */
+		rte_memcpy(ipv6_route.ip,
+			ip_addr->ip.v6.s6_addr, sizeof(ipv6_route.ip));
+		ipv6_route.depth = prefix_len;
+		ipv6_route.nexthop = fib_id;
+		ret = gk_lpm_add_ipv6_routes(
+			ltbl->lpm6, &ipv6_route, 1, ltbl);
+		if (ret < 0)
+			goto out;
+	} else
+		goto out;
+
+	new_fib->ref_cnt = 1;
+
+out:
+	return new_fib;
+}
+
+static int
+setup_neighbor_tbl(unsigned int socket_id, int identifier,
+	int ip_ver, int ht_size, struct neighbor_hash_table *neigh)
+{
+	int  ret;
+	char ht_name[64];
+	int key_len = ip_ver == ETHER_TYPE_IPv4 ?
+		sizeof(struct in_addr) : sizeof(struct in6_addr);
+
+	struct rte_hash_parameters neigh_hash_params = {
+		.entries = ht_size,
+		.key_len = key_len,
+		.hash_func = DEFAULT_HASH_FUNC,
+		.hash_func_init_val = 0,
+	};
+
+	ret = snprintf(ht_name, sizeof(ht_name),
+		"neighbor_hash_%u", identifier);
+	RTE_VERIFY(ret > 0 && ret < (int)sizeof(ht_name));
+
+	neigh->tbl_size = ht_size;
+
+	/* Setup the neighbor hash table. */
+	neigh_hash_params.name = ht_name;
+	neigh_hash_params.socket_id = socket_id;
+	neigh->hash_table = rte_hash_create(&neigh_hash_params);
+	if (neigh->hash_table == NULL) {
+		RTE_LOG(ERR, HASH,
+			"The GK block cannot create hash table for neighbor FIB!\n");
+
+		ret = -1;
+		goto out;
+	}
+
+	/* Setup the Ethernet header cache table. */
+	neigh->cache_tbl = rte_calloc(NULL,
+		ht_size, sizeof(struct ether_cache), 0);
+	if (neigh->cache_tbl == NULL) {
+		RTE_LOG(ERR, MALLOC,
+			"The GK block cannot create Ethernet header cache table\n");
+
+		ret = -1;
+		goto neigh_hash;
+	}
+
+	ret = 0;
+	goto out;
+
+neigh_hash:
+	rte_hash_free(neigh->hash_table);
+	neigh->hash_table = NULL;
+out:
+	return ret;
+}
+
+static int
+init_fib_tbl(struct gk_config *gk_conf)
+{
+	int ret;
+	unsigned int i;
+	int ip_prefix_len;
+	struct ipaddr addr;
+	unsigned int socket_id = rte_lcore_to_socket_id(gk_conf->lcores[0]);
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+	struct gk_fib *fib[4];
+
+	for (i = 0; i < GK_MAX_NUM_FIB_ENTRIES; i++) {
+		ltbl->fib_tbl[i].action = GK_FIB_MAX;
+		rte_spinlock_init(&ltbl->fib_tbl[i].lock);
+		rte_atomic16_init(&ltbl->fib_tbl[i].num_updated_instances);
+
+		ltbl->fib_tbl6[i].action = GK_FIB_MAX;
+		rte_spinlock_init(&ltbl->fib_tbl6[i].lock);
+		rte_atomic16_init(&ltbl->fib_tbl6[i].num_updated_instances);
+	}
+
+	/* Set up the FIB entry for the IPv4 front network prefix. */
+	addr.proto = ETHER_TYPE_IPv4;
+	rte_memcpy(&addr.ip.v4, &gk_conf->net->front.ip4_addr,
+		sizeof(addr.ip.v4));
+	ip_prefix_len = gk_conf->net->front.ip4_prefix_len;
+
+	fib[0] = add_prefix_fib(&addr, ip_prefix_len, gk_conf);
+	if (fib[0] == NULL) {
+		ret = -1;
+		goto out;
+	}
+
+	fib[0]->action = GK_FWD_NEIGHBOR_FRONT_NET;
+
+	ret = setup_neighbor_tbl(socket_id, 0, ETHER_TYPE_IPv4,
+		(1 << (32 - ip_prefix_len)), &fib[0]->u.neigh);
+
+	rte_spinlock_unlock(&fib[0]->lock);
```

Comment:
(1) The lock is from ``get_empty_fib_id()``, that is when allocating a new FIB entry, we first need to lock this entry, we won't release the lock until it is filled up.

(2) True, I made ``get_empty_fib_id()`` general, so that it allows multiple writers. Do I need to assign the very first FIB entries without using this function? Since the code is at the initialization part, the very first entries must be empty ones.